### ISSUE TITLE
CSS flavor support?

### ISSUE THREAD

**Issue by lydell** [prettier/prettier#1784]
Right now, the WIP "CSS" formatter seems to be using Less? I'd like support for SCSS ([postcss-scss](https://github.com/postcss/postcss-scss)).  I guess people writing vanilla CSS would like syntax errors for anything preprocessor syntax that isn't valid CSS as well. Has there been any thoughts on this?

**Comment by vjeux** [comment]
I'd like to support raw css, less, scss and sass.

What I was thinking is to extract the source type from the filename (and add a filename option to the js api). Then it automatically choses the best parser for it.

Then the --parser option would just be a way to change flow/babylon (and probably in the future typescript/babylon).

**Comment by vjeux** [comment]
CSS-Modules seem to be regular CSS-enough so they just work out of the box :)

```css
:global(.otherClassName) {
  composes: className from "./style.css";
}

@value blue: #0c77f8;
@value red: #ff0000;
@value green: #aaf200;

/* import your colors... */
@value colors: "./colors.css";
@value blue, red, green from colors;

.button {
  color: blue;
  display: inline-block;
}
```

(from their docs) is printed as is.

**Comment by vjeux** [comment]
I'm confused about SCSS, it seems like it's an extension of CSS but that has the same syntax as less. All the SCSS examples I could find print well with postcss-less. Do you know of some that don't?

```css
@mixin rounded($amount) {
  -moz-border-radius: $amount;
  -webkit-border-radius: $amount;
  border-radius: $amount;
}

.box {
  border: 3px solid #777;
  @include rounded(0.5em);
}

// These comments are only one line long each.
// They won't appear in the CSS output,
// since they use the single-line comment syntax.
a { color: green; }

@import "themes/dark";
@import "font.sass";
```

**Comment by SimenB** [comment]
I'd love to see support for Stylus, but I'm guessing that's blocked on a postcss parser for it?

**Comment by vjeux** [comment]
@SimenB looks like they have a proper parser: https://github.com/stylus/stylus/blob/dev/lib/parser.js, may not be too hard. But it's unfortunate that it's not based on postcss as we need to re-do the entire printer for it.

**Comment by vjeux** [comment]
For the error messages, we've opted to parse as much as possible with prettier. For babylon and flow, we enable all the options even low stage ones. It would be nice to provide good error messages for invalid syntax from prettier but it's not a requirement. If postcss-less is a superset of css, scss and less, I'm totally fine with it.

**Comment by vjeux** [comment]
Looks like postcss-sass is not working correctly. Only a single commit and: "Not all Sass syntax supported. Parser under development." https://github.com/aleshaoleg/postcss-sass

**Comment by SimenB** [comment]
If you circle around to doing stylus, this makes it _parse_, but not print.

```diff
diff --git i/src/parser.js w/src/parser.js
index 5cf4f0c8..692d8f02 100644
--- i/src/parser.js
+++ w/src/parser.js
@@ -16,6 +16,8 @@ function parse(text, opts) {
     parseFunction = parseWithTypeScript;
   } else if (opts.parser === "postcss") {
     parseFunction = parseWithPostCSS;
+  } else if (opts.parser === "stylus") {
+    parseFunction = parseWithStylus;
   } else {
     parseFunction = parseWithBabylon;
   }
@@ -338,6 +340,14 @@ function parseNestedCSS(node) {
   return node;
 }
 
+function addStylusType(node) {
+  node.type = `css-${node.nodeName}`;
+
+  if (node.nodes && node.nodes.length) {
+    node.nodes.forEach(addStylusType);
+  }
+}
+
 function parseWithPostCSS(text) {
   const r = require;
   const parser = r("postcss-less");
@@ -354,4 +364,22 @@ function parseWithPostCSS(text) {
   }
 }
 
+function parseWithStylus(text) {
+  const r = require;
+  const stylus = r("stylus");
+  try {
+    const result = new stylus.Parser(text).parse();
+    addStylusType(result);
+    return result;
+    // TODO: Is this needed?
+    const parsedResult = parseNestedCSS(result);
+    return parsedResult;
+  } catch (e) {
+    if (typeof e.lineno !== "number") {
+      throw e;
+    }
+    throw createError(e.name + " " + e.reason, e.lineno, e.column);
+  }
+}
+
 module.exports = { parse };
```

It ended up recursing infinitely using `addTypePrefix`. Can open up separate issue? I doubt I'll have time to work on it myself, though.

**Comment by vjeux** [comment]
@SimenB parseNestedCSS is needed because postcss is "modular" where it just parses the overall shape, then you need to use specialized parsers for values, selectors and media queries. In stylus, you should just be able to parse it straight away.

**Comment by lydell** [comment]
@vjeux 

Good point about that we're already enabling all kinds of JS features (JSX, type annotations, experimental syntax) even if the user don't use all of them. So I guess CSS will be just fine the same way.

Less and SCSS have different interpolation syntax. I tried postcss-less and postcss-scss and neither of them could parse the other one's interpolation syntax:

```less
// Less
.@{foo} {color: black;}
```

```scss
// SCSS
.#{foo} {color: black;}
```

**Comment by vjeux** [comment]
Looks like we have 0 occurrence of @{ inside of Nuclide codebase. Do you have any #{ in your scss codebase?

**Comment by vjeux** [comment]
Would be great to add a giant less test with all the examples of their docs: http://lesscss.org/features/

**Comment by vjeux** [comment]
Less doesn't seem to parse nested properties the way scss does:

```css
a { margin: 0 { left: 10px; } }
```

**Comment by lydell** [comment]
I try to avoid interpolation in selectors. When grepping through lots of code I couldn't find any occurrence _(pats oneself on the shoulder.)_ I did find this one though:

```scss
@media #{$g-breakpoint-tiny} {}
```

postcss-less couldn't parse the above, but the below (and vice versa for postcss-scss):

```less
@media @{$g-breakpoint-tiny} {}
```

Grepping through some vendor code, I found that FontAwesome uses lots of selector interpolation:

```scss
.#{$fa-css-prefix}-glass:before { content: $fa-var-glass; }
```

This is the by far most common pattern where we're using SCSS interpolation:

```scss
a {height: calc(#{$foo} + 1);}
```

Both postcss-scss and postcss-less seem allow both `#{` and `@{` in that position. Not sure how postcss-value-parser (or whatever it is called) deals with it though.

**Comment by vjeux** [comment]
```css
a {height: calc(#{$foo} + 1);}
```

is printed as is right now :)

```css
@media @{$g-breakpoint-tiny} {}
.@{$fa-css-prefix}-glass:before { content: $fa-var-glass; }
```

(with `@{`) are printed just fine.

Hacky solution would be to `s/#{/@{/` before and `s/@{/#{/` after printing if there's any `#{` in the file...

So sad that nested properties won't work :( I couldn't find any other thing that were different looking at postcss-scss tests

**Comment by vjeux** [comment]
Can you look if `(margin|padding):.*{` has any hits?

**Comment by lydell** [comment]
`\w+:\s*\{` gave me these results:

```scss
    background: {
        size: auto 60%;
        position: bottom 2px left;
    }
```

```scss
    background: {
        image: url('img/bg-blurred.jpg');
        color: #1c1d18;
        repeat: no-repeat;
        size: 100%;
    }
```

```scss
    background: {
        image: url('data:image/svg+xml,%3C%3Fxml%20version%3D%221.0%22%20%3F%3E%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20100%2086.6%22%3E%3Cpolygon%20points%3D%220%2C0%20100%2C0%2050%2C86.6%22%20fill%3D%22%23cdc6c0%22%2F%3E%3C%2Fsvg%3E');
        size: auto 1ex;
        position: right 0.5em center;
    }
```

I usually avoid this shorthand syntax though because of stylelint/stylelint#2177.

**Comment by vjeux** [comment]
Okay, so looks like we're going to have to support both! Let's do it :)

**Comment by vjeux** [comment]
What we can do for now (since we're not going to support sass anytime soon) is to try one and then the other. We can use a quick regex for `\w+:\s*\{` or `#{` to check if it's likely an scss one. We're already doing this for typescript with or without jsx.

**Comment by vjeux** [comment]
https://github.com/prettier/prettier/pull/1790

**Comment by vjeux** [comment]
@lydell would be awesome if you could run prettier with `--parser postcss` and `--debug-check` on all your scss and see if anything comes up :)

**Comment by lydell** [comment]
@vjeux Here you go: https://gist.github.com/lydell/4aed77086bc1c0a16193e73380908c5f

**Comment by vjeux** [comment]
Awesome, thanks for the results. I'm glad there aren't a lot and only one is remaining :)

- [x] _glyphicons.scss #1799
- [x] font-face.scss #1797
- [x] hamburger.scss: this is actually an issue with the parser that parses it differently if there's a space in-between, but the result is correct for both.
- [ ] parens.scss: Good catch. I need to run now but need to figure it out. I created #1803 to track it.

**Comment by lydell** [comment]
@vjeux Great!

- _glyphicons.scss: Unfortunately, the empty line above the comment is now unstable.
- hamburger.scss: So we need to open an issue somewhere?

Also, I skipped one directory of SCSS in my previous run, because the results were too messy. I'll try to go through them now.

**Comment by vjeux** [comment]
> the empty line above the comment is now unstable.

Yeah, we use location to figure out empty lines, if the locations are messed up, there's limited things we can do :(

> So we need to open an issue somewhere?

css-values-parser (with an s) would be the place. But I'm not sure it should be fixed, idk.

**Comment by lydell** [comment]
@vjeux So the only thing left in this issue is hamburger.scss. I now finally understand what's going on there. I thought all the time that it had to do with multi-line linear-gradient being unstable or something, but it's just the good old missing space before `(`. So, everything here is fixed now. 🎉

**Comment by phun-ky** [comment]
@SimenB just out of curiosity, did you fiddle a bit more with the support for stylus in prettier? :)

**Comment by evilebottnawi** [comment]
@phun-ky need `postcss-stylus` and a lot of tests

**Comment by phun-ky** [comment]
@evilebottnawi https://github.com/seaneking/poststylus ✔️ now where are those tests.. 🔍

**Comment by evilebottnawi** [comment]
@phun-ky PR welcome to add supports, need also tests in `prettier`, I'm supporting `css/less/scss/css-in-js` and it take many time, i don't have time for `stylus`, sorry

**Comment by SimenB** [comment]
@phun-ky Nope. I started adding support for prettier to ASTExplorer so it would be easier to handle, but got stuck on a bug in webpack, so I dropped it.

https://github.com/webpack/webpack/issues/4940
